README.rst

```diff
@@ -15,104 +15,171 @@ sqlitedict -- persistent ``dict``, backed-up by SQLite and pickle
+
+It's possible to use a custom (de)serializer, notably for JSON and for compression.
 
 .. code-block:: python
 
-  >>> import sqlitedict
-  >>> help(sqlitedict)
+    # use JSON instead of pickle
+    import json
+    mydict = SqliteDict("./my_db.sqlite", encode=json.dumps, decode=json.loads)
+
+    # apply zlib compression after pickling
+    import zlib, pickle, sqlite3
+
+    def my_encode(obj):
+        return sqlite3.Binary(zlib.compress(pickle.dumps(obj, pickle.HIGHEST_PROTOCOL)))
+    def my_decode(obj):
+        return pickle.loads(zlib.decompress(bytes(obj)))
+
+    mydict = SqliteDict("./my_db.sqlite", encode=my_encode, decode=my_decode)
+
+More
+----
 
-(but it's just ``dict`` with a commit, really).
+Functions are well documented, see docstrings directly in ``sqlitedict.py`` or call ``help(sqlitedict)``.
 
 **Beware**: because of Python semantics, ``sqlitedict`` cannot know when a mutable
-SqliteDict-backed entry was modified in RAM. For example, ``mydict.setdefault('new_key', []).append(1)``
-will leave ``mydict['new_key']`` equal to empty list, not ``[1]``. You'll need to
-explicitly assign the mutated object back to SqliteDict to achieve the same effect:
+SqliteDict-backed entry was modified in RAM. For example, ``db["123"]["name"] = "hello world"``
+will leave ``mydict["123"]`` object as it was, not changing the name. You'll need to
+explicitly assign the mutated object back to SqliteDict:
 
 .. code-block:: python
 
-  >>> val = mydict.get('new_key', [])
-  >>> val.append(1)  # sqlite DB not updated here!
-  >>> mydict['new_key'] = val  # now updated
+    val = db.get("123", {})
+    val["name"] = "hello world" # sqlite DB not updated here!
+    db["123"] = val  # now updated
 
+    db.commit() # remember to commit (or set autocommit)
 
-For developers
---------------
+Features
+========
+
+* Values can be **any picklable objects** (uses ``cPickle`` with the highest protocol).
+* Support for **multiple tables** (=dicts) living in the same database file.
+* Support for **access from multiple threads** to the same connection (needed by e.g. Pyro).
+  Vanilla sqlite3 gives you ``ProgrammingError: SQLite objects created in a thread can
+  only be used in that same thread.``
+
+  Concurrent requests are still serialized internally, so this "multithreaded support"
+  **doesn't** give you any performance benefits. It is a work-around for sqlite limitations in Python.
+
+* Support for **custom serialization or compression**:
+
+Performance
+===========
+
+* sqlite is efficient and can work effectively with large databases (multi gigabytes), not limited by memory.
+* sqlitedict is mostly a thin wrapper around sqlite, conserving efficiency.
+* ``items()`` ``keys()`` ``values()`` are iterating one by one, ``len()`` is calling sqlite to count rows.
+* For better performance, write objects in batch and ``commit()`` once.
+* When using pickle, make sure cPickle is installed (pip install cPickle).
```

Comment:
Why? AFAIK installed by default on Python 2, and not needed at all in Python 3.

Reply:
@morotti Ping on this item.

Reply:
Double checked. cPickled replaced pickle in python 3, no need to install, indeed. Will remove the line.